contributions/permissions.py

```diff
@@ -18,3 +20,25 @@ def has_permission(self, request, view):
             return False
             
         return self._is_member(request.user, chama_id)
+
+    def _check_permissions(self, contribution):
+        # Check if user has permission to update/delete a contribution
+        try:
+            membership = Membership.objects.get(user=self.request.user, chama=contribution.chama)
+        except Membership.DoesNotExist:
+            raise PermissionDenied("You are not a member of this Chama.")
+        
+        if membership.status != Membership.Status.ACTIVE:
+            raise PermissionDenied("Only users with ACTIVE membership can perform this action.")
+        
+        return membership.role == Membership.Role.ADMIN, membership
+    
+    def partial_update_permission(self, request, contribution):
+        # Memebrs can only delete their own PENDING or FAILED contributions
+        if contribution.member.user != request.user:
+            raise PermissionDenied("You can only delete your own contributions.")
+        
+        if contribution.status not in [Contribution.Status.PENDING, Contribution.Status.FAILED]:
+            raise PermissionDenied("Only pending or rejected contributions can be deleted.")
+        
+        return True
```

Comment:
The current logic is a bit misleading and not very clean, it only mentions updates, yet it’s also used for deletions. A simpler and clearer approach would be to use `has_object_permission` and handle method-specific checks for `PUT`, `PATCH`, and `DELETE` within that method.